detect/gcp/gce.go

```diff
@@ -0,0 +1,116 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcp
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	"cloud.google.com/go/compute/metadata"
+
+	"go.opentelemetry.io/otel/api/kv"
+	"go.opentelemetry.io/otel/api/standard"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+// GCE collects resource information of GCE computing instances
+type GCE struct{}
+
+// Detect detects associated resources when running on GCE hosts.
+func (gce *GCE) Detect(ctx context.Context) (*resource.Resource, error) {
+	if !metadata.OnGCE() {
+		return nil, nil
+	}
+
+	labels := []kv.KeyValue{
+		standard.CloudProviderGCP,
+		standard.CloudRegionKey.String(""),
+	}
+
+	var errInfo []string
+
+	projectID, err := metadata.ProjectID()
+	logError(err)
+	if projectID != "" {
+		labels = append(labels, standard.CloudAccountIDKey.String(projectID))
+	}
+	if err != nil {
+		errInfo = append(errInfo, err.Error())
+	}
```

Comment:
There is duplicate error handling here. They should be consolidated.

My recommendation would be to change the `logError` to something that validates if the error is not `nil` or for missing resources. Possibly replacing with `problem`:

```
func problem(err error) bool {
	if err == nil {
		return false
	}
	if _, undefined := err.(metadata.NotDefinedError); !undefined {
		return false
	}
	return true
}
```

And then this could be updated to:

```suggestion
	if projectID, err := metadata.ProjectID(); problem(err) {
		errInfo = append(errInfo, err.Error())
	} else if projectID != "" {
		labels = append(labels, standard.CloudAccountIDKey.String(projectID))
	}
```

This suggestion could likely use some cleanup.